joomla/joomla-cms
# [showon] Allow "not in" condition and simplify php code

Pull Request for Improvement.

### Summary of Changes

This PR does two things:
- Add a new "not in" condition to shown, we can use this condition by making use of the `!:` operator, insteadd of the normal `:`. This is particular useful if you want to put, for instance: field A should show on when fieldB value is not empty, ie, `showon="fieldB!:"`, which is not possible without this PR
Example: 
```php
// - Current
// Field A with value in 1 OR field B with value IN 1, 2 or 3
fieldname:1[OR]fieldB:1,2,3

// - NEW
// Field A with value in 1 OR field B with value NOT IN 1, 2 or 3
fieldname:1[OR]fieldB!:1,2,3
```
- Normalizes showon php code across the core. Remvoe a lot of duplicated code.

### Testing Instructions

1. Apply patch
2. Confirm current fields with showon behaviour continue to work fine
3. Change [this line](https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_config/model/form/application.xml#L553) line in com_config xml so the showon condition is `mailonline:1[AND]mailer!:smtp`.
Now check this field will appear when the field `mailonline` is set to yes and the mailer is different than smtp.
4. Make any other tests you can remember

### Documentation Changes Required

Add this new operator to https://docs.joomla.org/Form_field#Showon
